### PR TITLE
Scrum 73 implement pytest framework

### PR DESCRIPTION
**Pull request by maxmir20**
Added:
- `pytest.ini` Pytest config file
- `test_media_root` directory to hold dummy resume data for testing and file uploads/generation
- `test_models.py` for Resume `get_text` method
- `test_job_postings.py `for LinkedInPosting `format_url `method

Changed:
- Refactored `LinkedInPosting.format_url` method to improve readability and cover testing edge cases
- Modified gitignore to not include test_media
- Added pytest packages to `requirements.txt`

### CODE REVIEW

File: backend/tailor/domain/job_posting/linkedin.py [hunk 1]
```diff
@@ -15,22 +17,22 @@ def format_url(self, url):
             parsed_url = urlparse(url)
 
             # When URL is of the format https://www.linkedin.com/jobs/view/4259447405/?alternateChannel=search
-            full_path = urljoin(url, parsed_url.path)
-            if JOB_POSTING_PATH in full_path:
-                return full_path
+            if re.search(JOB_POSTING_REGEX, url):
```
Comment (maxmir20) [inline backend/tailor/domain/job_posting/linkedin.py:20]: this will prevent urls without a job id like `https://www.linkedin.com/jobs/view/` from passing

File: backend/tailor/tests/test_job_posting.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+from tailor.domain.job_posting import LinkedInPosting
+from tailor.exceptions import ParsingError
+
+
+class TestLinkedInPosting:
+
+    class TestFormatUrl:
+
+        @pytest.mark.parametrize("valid_url", [
```
Comment (maxmir20) [inline backend/tailor/tests/test_job_posting.py:11]: For more on how pytest parameterizing works, see here: https://docs.pytest.org/en/latest/example/parametrize.html

File: backend/tailor/tests/test_job_posting.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            "www.indeed.com",
+            "https://www.linkedin.com/",
+            "https://www.linkedin.com/jobs/view/",
+            "www.linkedin.com/jobs/view/4272165290/",
+        ])
+        def test_format_invalid_urls(self, invalid_url):
+            with pytest.raises(ParsingError):
+                LinkedInPosting(invalid_url)
+
+
```
Comment (maxmir20) [inline backend/tailor/tests/test_job_posting.py:30]: I am not testing `get_text` for now because this will fail if the linkedin job posting is expired, so we would need a way to dynamically fetch a new one

File: backend/tailor/tests/test_models.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+
+@pytest.fixture
+def user_data(db) -> User:
+    return User.objects.create_user("Test_User")
+
+
+@pytest.fixture
+def resume_data( request, db, user_data):
+    filename = request.param
```
Comment (maxmir20) [inline backend/tailor/tests/test_models.py:27]: request is a built in fixture to pytest that allows you to pass in parameters to the pytest fixtures, allowing us to create different types of resume_data. (e.g. pdf, docx). This is being done in the `TestResume.test_create_resume` method on ln40

See:
https://tech.serhatteker.com/post/2024-01/dynamically-passing-parameters-to-pytest-fixtures/
https://pytest-with-eric.com/fixtures/built-in/pytest-request-fixture/

File: backend/.gitignore [hunk 5]
```diff
@@ -25,5 +25,10 @@ ENV/
 
 # Files
 uploads/
+!backend/tailor/tests/test_media_root/uploads/
```
Comment (maxmir20) [inline backend/.gitignore:28]: Added some exceptions so we can add test data to our repo (not currently working)

File: backend/.gitignore [hunk 6]
```diff
@@ -24,6 +24,7 @@ venv/
 ENV/
 
 # Files
-uploads/
-*.pdf
-*.docx
+uploads/*
+!tailor/tests/test_media_root/uploads/*
```
Comment (maxmir20) [inline backend/.gitignore:28]: adds exception for test data

Comment (arvchandra) [reply backend/.gitignore:28]: Could this be possible without the exception?

As in, could we just gitignore the specific uploads folder: `backend/uploads`

File: backend/tailor/tests/test_models.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import shutil
+from pathlib import Path
+import pytest
+from django.conf import settings
+from django.contrib.auth.models import User
+from tailor.models import Resume, TailoredResume
+
+
+@pytest.fixture(autouse=True)
```
[the remark this answers is not in the document]
Comment (arvchandra) [reply backend/tailor/tests/test_models.py:9]: cool

**Review by arvchandra — APPROVED** [review]
Good job man. Just fix the end-of-file newlines on some of the files and you're good to go.
